### PR TITLE
More interface ring options

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:**

## Feature Details
Adds support for limiting the span and adjusting the starting angle for rings drawn with an interface.

Both values are given in degrees.
A span angle of less than 360 will accordingly scale down the value that is set for this ring in the given interface information. A span angle of 180 with a value of 0.5 will draw a quarter circle, instead of the half circle it would draw with a span angle of 360.
~~I'll have to check how the start angle part works.~~

## Usage Examples
```html
ring "my arc"
    "span angle" 90
    "start angle" 45
```

### CODE REVIEW

File: source/Interface.cpp [hunk 1]
```diff
@@ -654,7 +684,11 @@ void Interface::BarElement::Draw(const Rectangle &rect, const Information &info,
 		if(!rect.Width() || !rect.Height())
 			return;
 
-		RingShader::Draw(rect.Center(), .5 * rect.Width(), width, value, *color, segments);
+
+		double fraction = value * spanAngle / 360.;
+
+		RingShader::Draw(rect.Center(), .5 * rect.Width(), width, fraction, *color, segments, startAngle);
+
```
Comment (warp-core) [inline source/Interface.cpp:691]:
```suggestion
		RingShader::Draw(rect.Center(), .5 * rect.Width(), width, fraction, *color, segments, startAngle);
```

File: source/Interface.cpp [hunk 2]
```diff
@@ -628,9 +628,39 @@ bool Interface::BarElement::ParseLine(const DataNode &node)
...
+	}
+	if(startAngle > 360.)
+	{
+		startAngle = 360.;
+		warning = "Warning: \"start angle\" may not exceed 360. Using 360 instead:";
+	}
+
+	if(!warning.empty())
+		node.PrintTrace(warning);
+
```
Comment (warp-core) [inline source/Interface.cpp:663]:
```suggestion
	spanAngle = max(0., spanAngle);
	spanAngle = spanAngle % 360.;
	startAngle = max(0., startAngle);
	startAngle = startAngle % 360.;

```

File: source/Interface.cpp [hunk 3]
```diff
@@ -628,9 +628,39 @@ bool Interface::BarElement::ParseLine(const DataNode &node)
...
 		width = node.Value(1);
+	else if(node.Token(0) == "span angle" && node.Size() >= 2)
+		spanAngle = node.Value(1);
+	else if(node.Token(0) == "start angle" && node.Size() >= 2)
+		startAngle = node.Value(1);
 	else
 		return false;
 
+	// Some input validation.
+	string warning;
```
Comment (warp-core) [inline source/Interface.cpp:639]:
```suggestion
```

File: source/Interface.cpp [hunk 4]
```diff
@@ -616,6 +616,12 @@ Interface::BarElement::BarElement(const DataNode &node, const Point &globalAncho
 	// Fill in a default color if none is specified.
 	if(!color)
 		color = GameData::Colors().Get("active");
+
+	// Make sure these values are within the allowed ranges.
+	spanAngle = max(0., spanAngle);
+	spanAngle = min(360., spanAngle);
+	startAngle = max(0., startAngle);
+	startAngle = min(360., startAngle);
```
Comment (quyykk) [inline source/Interface.cpp:624]: Do this when parsing directly, so the variable never has an invalid value

File: source/Interface.cpp [hunk 5]
```diff
@@ -628,6 +634,10 @@ bool Interface::BarElement::ParseLine(const DataNode &node)
 		color = GameData::Colors().Get(node.Token(1));
 	else if(node.Token(0) == "size" && node.Size() >= 2)
 		width = node.Value(1);
+	else if(node.Token(0) == "span angle" && node.Size() >= 2)
+		spanAngle = node.Value(1);
+	else if(node.Token(0) == "start angle" && node.Size() >= 2)
+		startAngle = node.Value(1);
```
Comment (warp-core) [inline source/Interface.cpp:640]:
```suggestion
	else if(node.Token(0) == "span angle" && node.Size() >= 2)
		spanAngle = max(0., min(360., node.Value(1)));
	else if(node.Token(0) == "start angle" && node.Size() >= 2)
		startAngle = max(0., min(360., node.Value(1)));
```

File: source/Interface.cpp (hunk 4 above)
Comment (warp-core) [inline source/Interface.cpp:624]:
```suggestion
```

**Review by quyykk — APPROVED** [review]

**Pull request merged**